Add displayItems(Collection<Holder>) helper to CreativeModeTab.Builder for use with DeferredRegisters

ChiefArug opened the pull request:
A few times in the discord I've found people constructing elaborate registry helper systems (well, a method and a list) to be able to add them all to a creative tab later, unaware that DeferredRegister keeps a collection of all of its items. This adds a method to the creative tab builder that uses that collection to add all of a mods items to a creative tab.

The items will be added in the same order they are added to the deferred register as it is backed by a LinkedHashMap.

ChiefArug committed:
Add displayItems(DeferredRegister<Item>) helper to CreativeModeTab.Builder to simplify adding all of a mods items to a creative tab.

Spinoscythe commented:
You can do this for `DeferredRegister.Blocks` too right?

ChiefArug commented:
A method for blocks could be added as well, however it would need to filter for blocks that have items and I generally don't see the usecase for only adding blocks to a tab without adding items (because you can only make one call to displayItems in a builder you would need to iterate the entries yourself anyway if you want to add more than just blocks).

ChiefArug commented:
Saying that, the parameter type could be changed to `DeferredRegister<? extends ItemLike>` which would do blocks and items (and any modded ItemLikes), and filtering added for if the item returned by ItemLike is `== Items.AIR`.

Spinoscythe commented:
> Saying that, the parameter type could be changed to `DeferredRegister<? extends ItemLike>` which would do blocks and items (and any modded ItemLikes), and filtering added for if the item returned by ItemLike is `== Items.AIR`.

Good point. If possible, I think implementing this is better than hardcoding `DeferredRegister<Item>`

ChiefArug committed:
Change generic to ? extends ItemLike
This lets it accept Item or Block DRs.

Also add filtering for AIR, as if that is not done it will cause errors later on (in particular if your DR<Block> contains blocks that do not have an item form).

robotgryphon commented on patches/net/minecraft/world/item/CreativeModeTab.java.patch:
```diff
@@ -230,6 +230,20 @@
 +        }
++
++        /**
++         * Helper to set this tabs contents to everything in the supplied {@link net.neoforged.neoforge.registries.DeferredRegister}.
```
Wording: This does not set tab contents, but rather appends/adds more entries from the DR.

ChiefArug replied:
It doesn't append. displayItems can only be called once as it sets a field to the value passed into it. See the last line of the javadoc.*

Technically the build tab contents event can add more, but that is in relation to the contents set here.

*with the exception of one mod I found mixing into displayItems and making it append.

Shadows-of-Fire commented on patches/net/minecraft/world/item/CreativeModeTab.java.patch:
```diff
@@ -230,6 +230,20 @@
++         * Entries added through this method are filtered out if {@link ItemLike#asItem()} returns {@link Items.AIR}, to account for blocks that do not have items.
++         * Note that like the vanilla method this overrides any other calls for this tab to {@link #displayItems}.
++         */
++        public CreativeModeTab.Builder displayItems(net.neoforged.neoforge.registries.DeferredRegister<? extends ItemLike> register) {
```
We should not be binding DR into vanilla systems like this. If you want an easier method of supplying this data to `CreativeModeTab$Builder`, it should take an output of a DR, but not a DR directly.

You should be able to have it take a `Collection<Holder<? extends ItemLike>>` or a similar container.

Shadows-of-Fire replied:
Though, unfortunately, due to the return signature of `getEntries`, it will need to be
`Collection<? extends Holder<? extends ItemLike>> entries`

ChiefArug replied:
I don't like the idea of not having DR there directly and that generic is kinda ugly (makes it a lot less obvious you can easily pass a DR in), but I can understand not wanting to bind DR to vanilla like this.

Shadows-of-Fire replied:
You can _suggest_ using `DeferredRegister#getEntries()` in the javadocs for the method, since it should be one of (if not the most) common cases.

ApexModder commented on patches/net/minecraft/world/item/CreativeModeTab.java.patch:
```diff
@@ -230,6 +230,20 @@
++            return this.displayItems((p, o) -> register.getEntries().stream()
++                    .map(net.neoforged.neoforge.registries.DeferredHolder::get)
++                    .map(ItemLike::asItem)
++                    .filter(i -> i != Items.AIR)
```
Since we support modded items being disabled via feature flags, there should maybe be an additional filter to only append enabled items.

Should be something like so iirc
`.filter(item -> item.isEnabled(p.enabledFeatures()))`

ChiefArug replied:
Oh yeah, I forgot that was a thing.

ChiefArug committed:
Migrate to accepting a Collection<? extends Holder<? extends ItemLike>>.
Add filtering for disabled items.
